news/_posts/2020-02-04-the-new-prestashop-test-framework.md

```diff
@@ -0,0 +1,95 @@
+- Selenium itself is not very stable: the test campaign could be run 3 or 4 times in a row on Travis and have different results, with different tests failing. Sometimes it couldn’t find a selector, or timed out when waiting for something… it was very difficult and time-consuming to make the tests more robust.
+- The campaign was too complex: it was trying to test everything at once. The campaign would test all the functionalities in a page, but also doing some [functional testing](https://devdocs.prestashop.com/1.7/testing/how-to-create-your-own-web-acceptance-tests/#functional-tests) (verifying some functionalities work well together) and some [end-to-end testing](https://devdocs.prestashop.com/1.7/testing/how-to-create-your-own-web-acceptance-tests/#end-to-end-tests-soon) (running through the application like a real user and testing a bunch of functionalities at once). The result was a very complex test suite, very brittle (end-to-end tests are really fragile by design) and a nightmare to maintain and fix.
+
+There were some concerns with the code too: although it was working, there was no real framework behind it. No separation of concerns, no proper use of inheritance, no distinction between the test logic and the page logic. It was very complicated to find the source of a problem sometimes, and with more than 8000 test cases it frequently was not possible to fix it.
+
+All these reasons made it clear that it was time to drop the whole thing and start over.
+
+
+## What is coming
+
+The QA team decided to work on the three following points.
+
+### Automation tool
+
+It was already decided to drop Selenium, but what would be a good replacement? There are a lot of options there, including SaaS solutions (Cypress, QAWolf, etc). After reviewing a few possibilities, the best choice was to go for [Puppeteer](https://github.com/puppeteer/puppeteer).
+
+Puppeteer is a Node library which provides a nice API to control Chrome. It can automate pretty much everything you do manually. It’s maintained by Google, which is always a good sign.
+It’s also more robust and more stable in headless (which was a killer feature for us) since it was designed with headless in mind.
+
+
+### New framework
+
+The aim was also to create a more robust way to create and maintain tests to avoid the problems we had with the last implementation. Since Puppeteer is “just” a library, the QA team would have to create a whole framework around it, and that was a good opportunity to take a look at the [Page Object Pattern](https://martinfowler.com/bliki/PageObject.html). This pattern makes you separate the test logic from the page logic.
```

Comment:
```suggestion
As mentioned, we wanted to use this opportunity to create a more robust way to create and maintain tests and avoid the issues encountered with the Selenium implementation. Since Puppeteer is just a component, QA team will create a complete framework around it, and we also introduced the use of [Page Object Pattern](https://martinfowler.com/bliki/PageObject.html). This pattern ensures you separate the test logic from the page logic.
```